xla/service/cpu/onednn_matmul_rewriter.cc

```diff
@@ -178,41 +151,155 @@ auto GELUActivation(HloInstruction* instr, HloInstruction** src) {
+  //    1. rank(oeprand) <= rank(dot) and
+  //    2. Starting from the last dim in backward direction, the dimension
+  //       size of operand is either 1 or same to dot.
+  auto operand_dims = operand->shape().dimensions();
+  auto dot_dims = dot->shape().dimensions();
+  if (operand_dims.size() <= dot_dims.size()) {
+    int operand_idx = operand_dims.size() - 1;
+    int dot_idx = dot_dims.size() - 1;
+    for (; operand_idx >= 0; --operand_idx, --dot_idx) {
+      if (operand_dims[operand_idx] != 1 &&
+          operand_dims[operand_idx] != dot_dims[dot_idx])
+        return false;
+    }
+    return true;
+  } else {
+    return false;
+  }
+}
+
+inline bool IsRowMajor(const Shape& shape) {
+  return LayoutUtil::IsMonotonicWithDim0Major(shape.layout());
+}
+
+// Whether the element type of instr is compatible with oneDNN kernels.
+// TODO(intel-tf): Restict compatible types based on instruction kind.
+inline bool CompatibleElementType(const HloInstruction* instr) {
+  PrimitiveType element_type = instr->shape().element_type();
+  return element_type == BF16 || element_type == F32;
+}
+
+// Type conversion from and to any of BF16 and FP32.
+// TODO(intel-tf): Support more types when enabled.
+template <typename Pattern>
+inline auto SupportedConvert(Pattern pattern) {
+  auto supported_convert = [](const HloInstruction* instr) -> bool {
+    return CompatibleElementType(instr) &&
+           CompatibleElementType(instr->operand(0));
+  };
+  return m::Convert(pattern).WithPredicate(supported_convert);
+}
+
+template <typename Pattern>
+inline auto SupportedConvert(HloInstruction** convert, Pattern pattern) {
+  auto supported_convert = [](const HloInstruction* instr) -> bool {
+    return CompatibleElementType(instr) &&
+           CompatibleElementType(instr->operand(0));
+  };
+  return m::Convert(convert, pattern).WithPredicate(supported_convert);
+}
+
+template <typename Pattern>
+inline auto BitcastWithReshapeSemantics(HloInstruction** bitcast,
+                                        Pattern pattern) {
+  // TODO(intel-tf): Add stronger condition that Bitcast does not have transpose
+  // semantics. Some of the HLO passes replaces Transpose with Bitcast. Here
+  // the layouts are checked to be rowmajor since the current pass runs after
+  // the layout assignment and oneDNN matmul is enabled for rowmajor layouts.
+  auto is_reshape = [](const HloInstruction* instr) -> bool {
+    if (instr) {
+      auto input_shape = instr->operand(0)->shape();
+      auto output_shape = instr->shape();
+      bool is_same_type = ShapeUtil::SameElementType(input_shape, output_shape);
+      bool has_equal_num_elems = ShapeUtil::ElementsIn(input_shape) ==
+                                 ShapeUtil::ElementsIn(output_shape);
+      bool has_rowmajor_layout =
+          IsRowMajor(input_shape) && IsRowMajor(output_shape);
+      return is_same_type && has_equal_num_elems && has_rowmajor_layout;
+    } else {
+      false;
+    }
```

Comment:
Same here. Early return may be a bit simpler.
```suggestion
    if (!instr) return false;
    auto input_shape = instr->operand(0)->shape();
    auto output_shape = instr->shape();
    bool is_same_type = ShapeUtil::SameElementType(input_shape, output_shape);
    bool has_equal_num_elems = ShapeUtil::ElementsIn(input_shape) ==
                                ShapeUtil::ElementsIn(output_shape);
    bool has_rowmajor_layout =
        IsRowMajor(input_shape) && IsRowMajor(output_shape);
    return is_same_type && has_equal_num_elems && has_rowmajor_layout;
```